File: api/plumber.R
# =============================================================================
# api/plumber.R  (Member 4)
# Purpose : R Plumber API — 8 endpoints serving the React frontend.
# Run     : Rscript api/run_api.R
# =============================================================================

# Plumber evaluates this file with api/ as the working directory.
PROJ_ROOT <- normalizePath("..", mustWork = TRUE)

BUNDLE_PATH  <- file.path(PROJ_ROOT, "analytics", "outputs", "model_bundle.rds")
CLEANED_PATH <- file.path(PROJ_ROOT, "analytics", "outputs", "cleaned_food_prices.csv")
EVAL_PATH    <- file.path(PROJ_ROOT, "analytics", "outputs", "evaluation_results.csv")
PFOOD_PATH   <- file.path(PROJ_ROOT, "analytics", "outputs", "per_food_evaluation.csv")
FEATURE_SRC  <- file.path(PROJ_ROOT, "analytics", "utils", "feature_builder.R")

suppressPackageStartupMessages({
  library(plumber)
  library(dplyr)
  library(readr)
  library(jsonlite)
  library(lubridate)
  library(caret)
  library(glmnet)
})

source(FEATURE_SRC)
source(file.path(PROJ_ROOT, "api", "middleware", "validators.R"))

# ---------------------------------------------------------------------------
# Load assets at startup (not per-request)
# ---------------------------------------------------------------------------
BUNDLE  <- NULL
PRICES  <- NULL
API_READY <- FALSE

load_assets <- function() {
  tryCatch({
    BUNDLE  <<- readRDS(BUNDLE_PATH)
    PRICES  <<- read_csv(CLEANED_PATH, show_col_types = FALSE) %>%
                  mutate(date = as.Date(date))
    API_READY <<- TRUE
    message("[API] Assets loaded successfully. Version: ", BUNDLE$version)
  }, error = function(e) {
    message("[API] WARNING: Could not load assets — ", conditionMessage(e))
    message("[API] Run the analytics pipeline first:")
    message("[API]   Rscript analytics/01_data_cleaning.R")
    message("[API]   ...through...")
    message("[API]   Rscript analytics/06_model_evaluation.R")
  })
}

load_assets()

# ---------------------------------------------------------------------------
# CORS filter (allow React dev server)
# ---------------------------------------------------------------------------
#* @filter cors
function(req, res) {
  res$setHeader("Access-Control-Allow-Origin",  "*")
  res$setHeader("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
  res$setHeader("Access-Control-Allow-Headers", "Content-Type")
  if (req$REQUEST_METHOD == "OPTIONS") {
    res$status <- 204
    return(list())
  }
  plumber::forward()
}

# ===========================================================================
# GET /api/health
# ===========================================================================
#* @get /api/health
#* @tag system
#* @serializer json
function(res) {
  list(
    status      = ifelse(API_READY, "ok", "degraded"),
    api_ready   = API_READY,
    model_version = if (!is.null(BUNDLE)) BUNDLE$version else NULL,
    data_cutoff   = if (!is.null(BUNDLE)) BUNDLE$data_cutoff else NULL,
    timestamp   = format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ")
  )
}

# ===========================================================================
# GET /api/foods
# ===========================================================================
#* @get /api/foods
#* @tag data
#* @serializer json
function(res) {
  if (!API_READY) return(json_error(res, 503, "API not ready — run the analytics pipeline first."))

  foods <- PRICES %>%
    filter(!is.na(price)) %>%
    distinct(commodity_id, commodity) %>%
    arrange(commodity) %>%
    rename(id = commodity_id, name = commodity)

  json_ok(foods)
}

# ===========================================================================
# GET /api/markets?commodity_id=...
# ===========================================================================
#* @get /api/markets
#* @param commodity_id:int The commodity ID to filter markets by
#* @tag data
#* @serializer json
function(commodity_id, res) {
  if (!API_READY) return(json_error(res, 503, "API not ready."))

  cid <- suppressWarnings(as.integer(commodity_id))
  if (is.na(cid)) return(json_error(res, 400, "commodity_id must be an integer."))
  if (!cid %in% BUNDLE$known_commodity_ids) {
    return(json_error(res, 404, sprintf("commodity_id %d not found.", cid)))
  }

  markets <- PRICES %>%
    filter(commodity_id == cid, !is.na(price)) %>%
    distinct(market_id, market, admin1, admin2) %>%
    arrange(market) %>%
    rename(id = market_id, name = market, province = admin1, district = admin2)

  json_ok(markets)
}

# ===========================================================================
# GET /api/prices/history?commodity_id=...&market_id=...
# ===========================================================================
#* @get /api/prices/history
#* @param commodity_id:int
#* @param market_id:int
#* @tag data
#* @serializer json
function(commodity_id, market_id, res) {
  if (!API_READY) return(json_error(res, 503, "API not ready."))

  v <- validate_ids(commodity_id, market_id, BUNDLE)
  if (!v$ok) return(json_error(res, v$status, v$error))

  history <- PRICES %>%
    filter(commodity_id == v$commodity_id, market_id == v$market_id,
           !is.na(price)) %>%
    select(date, price, monthly_change_pct, absolute_change_pct, spike_flag) %>%
    arrange(date) %>%
    mutate(
      date                = format(date, "%Y-%m"),
      price               = round(price, 2),
      monthly_change_pct  = round(monthly_change_pct, 4),
      absolute_change_pct = round(absolute_change_pct, 4)
    )

  if (nrow(history) == 0) {
    return(json_error(res, 404,
      sprintf("No price history for commodity_id=%d, market_id=%d.",
              v$commodity_id, v$market_id)))
  }

  meta <- list(
    commodity_id  = v$commodity_id,
    market_id     = v$market_id,
    currency      = "LKR",
    unit          = "KG",
    price_type    = "Retail",
    data_cutoff   = BUNDLE$data_cutoff,
    n_months      = nrow(history),
    first_month   = history$date[1],
    last_month    = history$date[nrow(history)]
  )

  json_ok(list(meta = meta, series = history))
}

# ===========================================================================
# GET /api/volatility/history?commodity_id=...&market_id=...
# ===========================================================================
#* @get /api/volatility/history
#* @param commodity_id:int
#* @param market_id:int
#* @tag data
#* @serializer json
function(commodity_id, market_id, res) {
  if (!API_READY) return(json_error(res, 503, "API not ready."))

  v <- validate_ids(commodity_id, market_id, BUNDLE)
  if (!v$ok) return(json_error(res, v$status, v$error))

  series <- PRICES %>%
    filter(commodity_id == v$commodity_id, market_id == v$market_id,
           !is.na(price)) %>%
    arrange(date)

  if (nrow(series) < 3) {
    return(json_error(res, 422,
      "Insufficient data to compute rolling volatility (need ≥ 3 months)."))
  }

  enriched <- build_features(series %>%
                               select(commodity_id, market_id, date, price))

  vol_series <- enriched %>%
    filter(!is.na(rolling_volatility_3m)) %>%
    select(date, rolling_volatility_3m, mean_absolute_change_3m) %>%
    mutate(
      date                    = format(date, "%Y-%m"),
      rolling_volatility_3m   = round(rolling_volatility_3m, 6),
      mean_absolute_change_3m = round(mean_absolute_change_3m, 4)
    )

  json_ok(list(
    meta   = list(commodity_id = v$commodity_id, market_id = v$market_id,
                  window_months = 3, data_cutoff = BUNDLE$data_cutoff),
    series = vol_series
  ))
}

# ===========================================================================
# POST /api/predictions
# Body: { "commodity_id": int, "market_id": int }
# ===========================================================================
#* @post /api/predictions
#* @tag predictions
#* @serializer json
function(req, res) {
  if (!API_READY) return(json_error(res, 503, "API not ready."))

  # Parse body
  body <- tryCatch(jsonlite::fromJSON(req$postBody), error = function(e) NULL)
  if (is.null(body)) {
    return(json_error(res, 400, "Request body must be valid JSON with commodity_id and market_id."))
  }

  v <- validate_ids(body$commodity_id, body$market_id, BUNDLE)
  if (!v$ok) return(json_error(res, v$status, v$error))

  # Retrieve latest history for this series
  series <- PRICES %>%
    filter(commodity_id == v$commodity_id, market_id == v$market_id,
           !is.na(price)) %>%
    arrange(date)

  if (nrow(series) < 4) {
    return(json_error(res, 422,
      "Insufficient price history. Need at least 4 consecutive months."))
  }

  # Build features using the canonical feature builder
  enriched <- build_features(series %>%
                               select(commodity_id, market_id, date, price))

  prep <- prepare_latest_row(enriched)
  if (!prep$ok) return(json_error(res, 422, prep$error))

  latest_row <- prep$row

  # Apply preprocessing
  impute_fn_api <- function(row) {
    for (col in names(BUNDLE$impute_medians)) {
      if (col %in% names(row) && is.na(row[[col]])) {
        row[[col]] <- BUNDLE$impute_medians[[col]]
      }
    }
    row
  }
  latest_imputed <- impute_fn_api(latest_row)

  X_pred <- tryCatch(
    predict(BUNDLE$preproc_caret,
            latest_imputed %>% select(all_of(BUNDLE$predictors))) %>% as.matrix(),
    error = function(e) NULL
  )
  if (is.null(X_pred)) {
    return(json_error(res, 500, "Feature preprocessing failed. Check predictor schema."))
  }

  # Generate continuous prediction (Ridge)
  ridge_pred <- pmax(as.vector(predict(BUNDLE$ridge_model, newx = X_pred)), 0)

  # Generate probability prediction (Logistic)
  logit_prob <- tryCatch(
    predict(BUNDLE$logit_model, newdata = as.data.frame(X_pred),
            type = "response"),
    error = function(e) NA_real_
  )

  # Warnings
  warnings <- character(0)
  if (isTRUE(latest_row$spike_flag)) {
    warnings <- c(warnings, "The most recent month was flagged as a potential price spike. Prediction may be less reliable.")
  }
  if (nrow(series) < 12) {
    warnings <- c(warnings, sprintf("This series has only %d months of history. Predictions are less reliable with short histories.", nrow(series)))
  }

  json_ok(
    list(
      commodity_id                = v$commodity_id,
      market_id                   = v$market_id,
      last_observed_month         = prep$last_observed_month,
      forecast_month              = prep$forecast_month,
      expected_absolute_change_pct = round(ridge_pred, 4),
      large_change_probability    = round(as.numeric(logit_prob), 4),
      large_change_threshold_pct  = BUNDLE$large_change_threshold_pct,
      model_version               = BUNDLE$version,
      data_cutoff                 = BUNDLE$data_cutoff,
      disclaimer = paste(
        "This prediction estimates movement size (up or down), not direction.",
        "It is based on historical patterns and not a live market signal.",
        sprintf("Data ends %s.", BUNDLE$data_cutoff)
      )
    ),
    warnings = warnings
  )
}

# ===========================================================================
# GET /api/models/performance
# ===========================================================================
#* @get /api/models/performance
#* @tag models
#* @serializer json
function(res) {
  if (!API_READY) return(json_error(res, 503, "API not ready."))
  if (is.null(BUNDLE$test_performance)) {
    return(json_error(res, 404,
      "Evaluation results not yet available. Run analytics/06_model_evaluation.R."))
  }

  perf <- BUNDLE$test_performance

  json_ok(list(
    continuous       = perf$continuous,
    binary           = perf$binary,
    per_commodity    = perf$per_commodity,
    test_period      = perf$test_period,
    n_test           = perf$n_test,
    model_version    = BUNDLE$version,
    large_change_threshold_pct = BUNDLE$large_change_threshold_pct,
    limitations = c(
      "Models trained on historical data up to December 2024.",
      "Performance may degrade for commodity-market pairs with sparse data.",
      "Ridge regression constrains negative predictions to 0.",
      "Observations within the same food-market series are not independent."
    )
  ))
}

# ===========================================================================
# GET /api/metadata
# ===========================================================================
#* @get /api/metadata
#* @tag system
#* @serializer json
function(res) {
  list(
    source       = "WFP Food Prices — Sri Lanka (Humanitarian Data Exchange)",
    source_url   = "https://data.humdata.org/dataset/wfp-food-prices-for-sri-lanka",
    currency     = "LKR",
    unit         = "KG",
    price_type   = "Retail",
    data_cutoff  = if (!is.null(BUNDLE)) BUNDLE$data_cutoff else "unknown",
    model_version = if (!is.null(BUNDLE)) BUNDLE$version else "unknown",
    api_version  = "1.0.0",
    disclaimer   = paste(
      "This is a historical prototype built on data ending September 2025.",
      "It is not a live forecasting service."
    )
  )
}

File: frontend/src/api/plumber.ts
import { z } from 'zod';

export const isLiveMode = import.meta.env.VITE_API_MODE !== 'mock';
const base = (import.meta.env.VITE_API_BASE_URL || '/api').replace(/\/$/, '');

// Plumber wraps scalar list fields in arrays; data-frame rows remain arrays.
const scalar = <T extends z.ZodType>(schema: T) =>
  z.union([schema, z.array(schema).length(1).transform(values => values[0])]);
const text = scalar(z.string());
const number = scalar(z.number());
const nullableNumber = scalar(z.number().nullable()).default(null);

const healthSchema = z.object({
  status: text,
  api_ready: scalar(z.boolean()),
  model_version: scalar(z.string().nullable()),
  data_cutoff: scalar(z.string().nullable()),
});
const metadataSchema = z.object({
  source: text, source_url: text, currency: text, unit: text,
  price_type: text, data_cutoff: text, model_version: text, disclaimer: text,
});
const foodSchema = z.object({ id: z.number(), name: z.string() });
const marketSchema = foodSchema.extend({ province: z.string(), district: z.string() });
const historySchema = z.object({
  meta: z.object({
    commodity_id: number, market_id: number, currency: text, unit: text,
    price_type: text, data_cutoff: text, n_months: number,
    first_month: text, last_month: text,
  }),
  series: z.array(z.object({
    date: z.string(), price: z.number(), monthly_change_pct: z.number().nullable().default(null),
    absolute_change_pct: z.number().nullable().default(null), spike_flag: z.boolean(),
  })),
});
const volatilitySchema = z.object({
  meta: z.object({ window_months: number, data_cutoff: text }),
  series: z.array(z.object({
    date: z.string(), rolling_volatility_3m: z.number(), mean_absolute_change_3m: z.number(),
  })),
});
const predictionSchema = z.object({
  commodity_id: number, market_id: number, last_observed_month: text,
  forecast_month: text, expected_absolute_change_pct: scalar(z.number().nonnegative()),
  large_change_probability: scalar(z.number().min(0).max(1).nullable()).default(null),
  large_change_threshold_pct: number, model_version: text, data_cutoff: text,
  disclaimer: text, warnings: z.array(z.string()).default([]),
});
const performanceSchema = z.object({
  continuous: z.array(z.object({ model: z.string(), mae: z.number(), rmse: z.number() })),
  binary: z.array(z.object({
    model: z.string(), brier_score: nullableNumber, precision: nullableNumber,
    recall: nullableNumber, f1: nullableNumber, accuracy: nullableNumber,
    tp: number, fp: number, fn: number, tn: number, prob_cutoff: number,
  })),
  per_commodity: z.array(z.object({
    commodity: z.string(), n: z.number(), mae_ridge: z.number(),
    mae_baseline: z.number(), brier_logit: z.number(), pct_large_true: z.number(),
  })),
  test_period: z.object({ start: text, end: text }), n_test: number,
  model_version: text, large_change_threshold_pct: number, limitations: z.array(z.string()),
});

const envelopeSchema = z.object({
  success: scalar(z.boolean()), data: z.unknown().optional(),
  error: text.optional(), warnings: z.array(z.string()).optional(),
});

async function request<T extends z.ZodType>(path: string, schema: T, init?: RequestInit): Promise<z.infer<T>> {
  let response: Response;
  try {
    response = await fetch(`${base}${path}`, {
      ...init, signal: AbortSignal.timeout(15_000),
      headers: { Accept: 'application/json', ...(init?.body ? { 'Content-Type': 'application/json' } : {}), ...init?.headers },
    });
  } catch {
    throw new Error('Could not reach the backend. Check that the R API is running on port 8000, then retry.');
  }
  const raw: unknown = await response.json().catch(() => null);
  const envelope = envelopeSchema.safeParse(raw);
  if (!response.ok || (envelope.success && !envelope.data.success)) {
    throw new Error(envelope.success && envelope.data.error
      ? envelope.data.error
      : `The backend could not complete this request (HTTP ${response.status}).`);
  }
  let payload = envelope.success ? envelope.data.data : raw;
  if (envelope.success && envelope.data.warnings && payload && typeof payload === 'object' && !Array.isArray(payload)) {
    payload = { ...payload, warnings: envelope.data.warnings };
  }
  const parsed = schema.safeParse(payload);
  if (!parsed.success) throw new Error('The backend returned an unexpected response format. Check the API version.');
  return parsed.data;
}

export interface SeriesSelection { commodity_id: number; market_id: number }
const query = (selection: SeriesSelection) => new URLSearchParams({
  commodity_id: String(selection.commodity_id), market_id: String(selection.market_id),
}).toString();

export const backend = {
  health: () => request('/health', healthSchema),
  metadata: () => request('/metadata', metadataSchema),
  foods: () => request('/foods', z.array(foodSchema)),
  markets: (commodityId: number) => request(`/markets?commodity_id=${commodityId}`, z.array(marketSchema)),
  history: (selection: SeriesSelection) => request(`/prices/history?${query(selection)}`, historySchema),
  volatility: (selection: SeriesSelection) => request(`/volatility/history?${query(selection)}`, volatilitySchema),
  predict: (selection: SeriesSelection) => request('/predictions', predictionSchema, {
    method: 'POST', body: JSON.stringify(selection),
  }),
  performance: () => request('/models/performance', performanceSchema),
};

export type Food = z.infer<typeof foodSchema>;
export type Market = z.infer<typeof marketSchema>;
